flocker/ca/_ca.py

```diff
@@ -51,138 +66,238 @@ def __init__(self, keypair):
         return False
 
     def __ne__(self, other):
         return not self.__eq__(other)
 
-    def selfSignedCACertificate(self, dn, request, serial, expiry, digest):
-        """
-        Sign a CertificateRequest with extensions for use as a CA certificate.
 
-        See
-        https://www.openssl.org/docs/apps/x509v3_config.html#Basic-Constraints
-        for further information.
+def create_certificate_authority(keypair, dn, request, serial, expiry, digest):
+    """
+    Sign a CertificateRequest with extensions for use as a CA certificate.
 
-        This code based on ``twisted.internet.ssl.KeyPair.signRequestObject``
+    See
+    https://www.openssl.org/docs/apps/x509v3_config.html#Basic-Constraints
+    for further information.
 
-        :param DistinguishedName dn: The ``DistinguishedName`` for the
-            certificate.
+    This code based on ``twisted.internet.ssl.KeyPair.signRequestObject``
 
-        :param CertificateRequest request: The signing request object.
+    :param KeyPair keypair: The private/public key pair.
 
-        :param int serial: The certificate serial number.
+    :param DistinguishedName dn: The ``DistinguishedName`` for the
+        certificate.
 
-        :param int expiry: Number of seconds from now until this certificate
-            should expire.
+    :param CertificateRequest request: The signing request object.
 
-        :param str digest: The digest algorithm to use.
-        """
-        req = request.original
-        cert = crypto.X509()
-        dn._copyInto(cert.get_issuer())
-        cert.set_subject(req.get_subject())
-        cert.set_pubkey(req.get_pubkey())
-        cert.gmtime_adj_notBefore(0)
-        cert.gmtime_adj_notAfter(expiry)
-        cert.set_serial_number(serial)
-        cert.add_extensions([
-            crypto.X509Extension("basicConstraints", True,
-                                 "CA:TRUE, pathlen:0"),
-            crypto.X509Extension("keyUsage", True,
-                                 "keyCertSign, cRLSign"),
-            crypto.X509Extension("subjectKeyIdentifier", False, "hash",
-                                 subject=cert),
-        ])
-        cert.add_extensions([
-            crypto.X509Extension(
-                "authorityKeyIdentifier", False,
-                "keyid:always", issuer=cert
-            )
-        ])
-        cert.sign(self.keypair.original, digest)
-        return Certificate(cert)
+    :param int serial: The certificate serial number.
 
+    :param int expiry: Number of seconds from now until this certificate
+        should expire.
 
-def generate_keypair():
+    :param str digest: The digest algorithm to use.
+    """
+    req = request.original
+    cert = crypto.X509()
+    dn._copyInto(cert.get_issuer())
+    cert.set_subject(req.get_subject())
+    cert.set_pubkey(req.get_pubkey())
+    cert.gmtime_adj_notBefore(0)
+    cert.gmtime_adj_notAfter(expiry)
+    cert.set_serial_number(serial)
+    cert.add_extensions([
```

Comment:
There is a comment explaining these; the link to ttps://www.openssl.org/docs/apps/x509v3_config.html#Basic-Constraints on line 82.